cmd/expiration-mailer/main.go

```diff
@@ -229,16 +231,92 @@ func (m *mailer) findExpiringCertificates() error {
 			m.log.Err(fmt.Sprintf("expiration-mailer: Error loading certificates: %s", err))
 			return err // fatal
 		}
-		if len(certs) > 0 {
-			processingStarted := m.clk.Now()
-			m.processCerts(certs)
-			m.stats.TimingDuration("Mailer.Expiration.ProcessingCertificatesLatency", time.Since(processingStarted), 1.0)
+
+		if len(certs) == 0 {
+			m.log.Info(fmt.Sprintf("expiration-mailer: Found 0 certificates that expire between %s and %s and had last nag >%s before expiry", left, right, expiresIn))
+			continue // nothing to do
 		}
+
+		certs = m.eliminateRenewedCerts(certs)
+
+		processingStarted := m.clk.Now()
+		m.processCerts(certs)
+		m.stats.TimingDuration("Mailer.Expiration.ProcessingCertificatesLatency", time.Since(processingStarted), 1.0)
 	}
 
 	return nil
 }
 
+// eliminateRenewedCerts returns a new slice with the already-renewed certificates removed.
+//
+// Structure of goroutines
+//
+//                /-- worker --\
+//  dispatcher ---+-- worker --+--- collector
```

Comment:
As I mentioned in person this is way too complex. A simpler approach would be to write a function that takes a single certificate serial and returns whether there's a newer certificate with the same fqdnSet. You can then call that from the `for _, cert := range certs {` loop inside `processCerts`.

This will be simpler to reason about and easier to test.